src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Dictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
```

Comment:
Did not we say we wanted to store WeakReference to cts, or somehow clean them up?